### DISCUSSION TITLE
[Proposal] Ternary conditional comparison operators

### DISCUSSION THREAD

**Discussion by slimCODE** [dotnet/csharplang#5629] [Language Ideas]
Given this naïve code:
```csharp
    var sortedItems = new[] {
        new { Tag = "One", IsHidden = false },
        new { Tag = "Four", IsHidden = false },
        new { Tag = "Seven", IsHidden = false },
        new { Tag = "Two", IsHidden = true },
        new { Tag = "Five", IsHidden = true }
    };

    string[] ExternalMethod(IEnumerable<string> tags, int firstHiddenIndex)
    {
        if (firstHiddenIndex < 0) throw new ArgumentException();
        return tags.Select((tag, index) => index >= firstHiddenIndex ? $"({tag})" : tag).ToArray();
    }

    var firstHiddenIndex = Array.FindIndex(sortedItems, item => item.IsHidden);
    if (firstHiddenIndex < 0)
    {
        firstHiddenIndex = sortedItems.Length;
    }

    var labels = ExternalMethod(sortedItems.Select(item => item.Tag), firstHiddenIndex);
```

It would be nice to be able to assign the result of a call to a variable if it meets a condition, but use a different value if it doesn't, like this:

```csharp
    var firstHiddenIndex = Array.FindIndex(sortedItems, item => item.IsHidden) ?!= -1 : sortedItems.Length;
```

Here, I'm introducing the `{left} ?!= {right} : {alt}` conditional operator, which means "Return the left value if it differs from the right value, otherwise return the alt value".

We would also have `?==`, `?>=`, `?<=`, `?>` and `?<` operators.

Those more in favor of pattern matching style could think of this instead:

```csharp
    var firstHiddenIndex = Array.FindIndex(sortedItems, item => item.IsHidden) when != -1 else sortedItems.Length;
```

**Comment by HaloFour** [comment]
> Those more in favor of pattern matching style could think of this instead.

Why not just use pattern matching here?

```cs
var firstHiddenIndex = Array.FindIndex(sortedItems, item => item.IsHidden) is not -1 and var index ? index : sortedItems.Length;
```

The example code with these ternary operators feels much less clear to me, and this kind of comparison not common enough to warrant a slew of new operators.

**Reply by slimCODE** [reply]
I find this redundant creation of a variable `index` to be too verbose and though the compiler is optimized, my mind tells me it's not efficient.

I also don't like the fact that the right side of the `and` is not a condition. It's not intuitive to me to write and expression there.

**Reply by HaloFour** [reply]
I don't think verbosity is a liability here.  Honestly I think the original code, with a normal `if` statement, is more clear than either other approach, although the `switch` approach isn't too bad and feels nice and FP to me:

```cs
var firstHiddenIndex = Array.FindIndex(sortedItems, item => item.IsHidden) switch {
    -1 => sortedItems.Length,
    var index => index
};
```

As for the efficiency, I fully expect that the compiler will emit a separate slot for a temporary local regardless of the approach taken as the assignment doesn't occur until after the expression is complete.  But I also expect that JIT will be able to optimize that away.

The right side of `and` is a pattern, and `var` is a pattern.  But I get that using patterns in this manner can also be confusing.
